Windows 10/11 virtual desktop switch announcements (take 3): use a combination of name change event and a dedicated event handler function to announce virtual desktop switches in Windows 10 Version 1903 and later

josephsl opened the pull request:
### Introduction
This is the third take on Windows 10/11 virtual desktop switch announcements, with prior efforts being #8259 and #11031 written by me, @MichaelDCurran, and @JCSTeh. The first version mostly came from Windows App Essentials add-on with improvements by Mick, with Jamie improving it to note improvements made in Windows 10 Version 1903 and later in the second version. Subsequently, it was noted that name change event comes from CSRSS/desktop object, with recent Windows App Essentials add-on releases including a dedicated CSRSS app module to announce virtual desktop names after a 100 millisecond delay. The next version of Windows App Essentials add-on (version 23.06) will include an approach similar to this pull request, with no delay in virtual desktop announcement except when switching between multiple desktops while focused on the desktop itself (the only thing different in the add-on version is that the add-on version will continue to use CSRSS app module; see below for details).

### Link to issue number:
Closes #5641 

### Summary of the issue:
IN Windows 10 and later, NVDA does not announce virtual desktop names when opening/closing/switching between them.

### Description of user facing changes
In Windows 10 May 2019 Update (Version 1903) and later, NVDA will announce virtual desktop names when opening, closing, and switching between virtual desktops.

### Description of development approach
Prior efforts used a combination of event handler and IAccessible object definition to announce virtual desktop names as follows:

* Define a "handle possible virtual desktop switch" function in event handler to announce virtual desktop names.
* Before handling gain focus event, see if virtual desktop was switched, and if yes, announce virtual desktop name after handling foregrond changes.
* Define an IAccessible object representing the current desktop, handling name change event coming from it when switching virtual desktops.
* The IAccessible object's name change event handler will queue virtual desktop switch function to be run after 250 millisecond delay. This handles virtual desktop switches when focused on the desktop.
* A virtual desktop name string is defined in event handler to keep track of last virtual desktop name, and if it differs, announce virtual desktop names. This works properly except in cases where consecutive virtual desktops have the same name, more so in Windows 11 where virtual desktop reordering is possible.

The third approach differs in that there is no definition of a dedicated IAccessible object for virtual desktop changes, as well as:

* Name change event from desktop object is handled as part of execute event routine.
* Virtual desktop name is still defined in event handler but only for the duration of virtual desktop switch announcement handling (up to 250 milliseconds).
* Virtual desktop name is cleared whenever name change occurs from CSRSS/desktop object so consecutive virtual desktops with identical names can be announced, including when reordered in Windows 11.

While the first two prior approaches could be improved to clear virtual desktop name field after handling virtual desktop switch announcements, the third take removes the need to edit IAccessible objects package. If the IAccessible object approach is more obust, then this PR can be edited to bring in Jamie's work and modernize it.

### Testing strategy:
Manual testing (prerequisite: Windows 10 1903 or later, disable Windows App Essentials add-on):

1. Open a new virtual desktop (Control+Windows+D).
2. Switch between virtual desktops (Control+Windows+left/right arrow keys).
3. Close a virtual desktop (Control+Windows+F4).
4. Open another virtual desktop.
5. Reanme the currently open virtual desktop (Task View/Windows+Tab, move to desktops list, select the current virtual desktop, press F2, type a new virtual desktop name, then press Enter).
6. Optional: switch to the immediate next or previous virtual desktop and rename it, giving it the same name as the one used before. Be sure virtual desktops next to each other have identical names.
7. Switch between virtual desktops.

### Known issues with pull request:
There could be up to 250 millisecond delay from switching virtual desktops to announcing virtual desktop names. Also, when closing virtual desktops, the name of the focused aplication/control might be announced prior to announcement of the new virtual desktop name, caused by focus event happening before name change from CSRSS/desktop object, and this is something that I'm sure people can live with.

### Change log entries:
New feature:

In Windows 10 May 2019 Update and later, NVDA will announce names of virtual desktops when opening, closing, changing, and reordering (Windows 11) virtual desktops. (#5641)

### Code Review Checklist:

<!--
This checklist is a reminder of things commonly forgotten in a new PR.
Authors, please do a self-review of this pull-request.
Check items to confirm you have thought about the relevance of the item.
Where items are missing (eg unit / system tests), please explain in the PR.
To check an item `- [ ]` becomes `- [x]`, note spacing.
You can also check the checkboxes after the PR is created.
A detailed explanation of this checklist is available here:
https://github.com/nvaccess/nvda/blob/master/devDocs/githubPullRequestTemplateExplanationAndExamples.md#code-review-checklist
-->

- [x] Pull Request description:
  - description is up to date
  - change log entries
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] API is compatible with existing add-ons.
- [x] Documentation:
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] Security precautions taken.

josephsl committed:
Event handler: update copyright header. Re #5641

josephsl committed:
Event handler/virtual desktops announcement: use a dedicated function to handle possible virtual desktop switches. Re #5641.

Original work by Jamie Teh (Mozilla): define a dedicated function to handle posible (and real) Windows 10/11 virtual desktop switche announcements. This function checks to make sure this is Widows 10 or later, and if yes, announces desktop name as defined in the event handler module, and clears this string once announcements are done. The actual event responsible for placing virtual desktop name in event handler is name change event from CSRSS (client/server runtime subsystem/Windows subsystem) process, specificlaly desktop object. Handling name change event will be done as part of handling gain focus event, specifically when doing pre-gain focus routine.

josephsl committed:
Event handler/virtual desktop switch announcements: detect name change event coming from CSRSS/desktop object and treat as a possible virtual desktop switch. Re #5641.

Original idea from Jamie Teh (Mozilla): in 'execute event' function, detect namechange coming from CSRSS/desktop object and treat it as a possible virtual desktop name change before handling focus events. If it turns out this is indeed a virtual desktop switch, set virtual desktop name, then queue handle possible virtual desktop switch routine (250 milliseconds) so desktop switches can be announced when opening/closing/switching virtual desktops while focused on the desktop.

josephsl committed:
Event handler/virtual desktop switch announcements: handle possible virtual desktop switches after handling new foreground event. Re #5641.

Credit: Jamie Teh (Mozilla): handle possible virtual desktop switch announcements after handling new foreground announcement in 'do pre gain focus' routine. This allows virtual desktop names to be announced between actual desktop switch and handling gain focus event, more noticeable when creating new virtual desktops.

josephsl committed:
Event handler/virtual desktop switch announcement: limit announcement to Windows 10 Version 1903 and later. Re #5641.

A note from Jamie Teh (Mozila): virtual desktop switch announcement works more effectively in Windows 10 Version 1903 (May 2019 Update) and later, therefore restrict virtual desktop announcement to that feature update and later.

josephsl committed:
Event handler/virtual desktop switch announcement and surrounding code: lint

josephsl committed:
Event handler/virtual desktop switch announcement: use formatted string literals when logging exceptions with execute event routine.

josephsl committed:
Event handler/virtual desktop switch announcement: add typing info for virtual desktop name field and handle possible desktop switch function. Re #5641

josephsl committed:
Event handler/virtual desktop switch announcement: use a dedicated variable to detect Windows 10 1903 and later. Re #5641.

In addition to placing winVersion module import at the top of the file, use a dedicated flag to indicate support for virtual desktop switch announcements in Windows 10 Version 1903 (May 2019 Update) and later.

XLTechie commented:
Thanks for this work @josephsl.

I've not tested or looked at this code yet, but would it be reasonable to create 
an extensionPoint here? I'm thinking an action, that notifies when a desktop is 
switched, and provides any relevant info (name, for example).

I'm not sure yet what the consumers of such might be, or if it would have any 
value internally (to this work or other), but it seems that this may not be so 
easily tracked by event capturing; and add-ons might want to know that the 
desktop has changed for some reason.

josephsl commented:
Hi, add-ons can listen to name change event and catch if it came from CSRSS (that was the approach taken in Windows App Essentials add-on in 2022 and earlier). At this point, I don’t see a use case for an action extension point, but perhaps it might be something to consider in the future (realistically speaking, as of May 2023, announcing virtual desktops is the key reason for using Windows App Essentials add-on; once this feature comes to NVDA Core, I can then start slowly transitioning the add-on to maintenance mode in Windows 10 environment). Thanks.

josephsl commented:
Hi, as a bonus, while the implementation step took about an hour (about two hours total if I include work done in the past to bring this to Windows App Essentials add-on), linting (and in one case, modernization) took more time as we are dealing with a code fragment that wasn’t touched until recently. To demonstrate the magnitude of this change, the code that logs an error when “executeEvent” fails uses string interpolation (modulo operator), something that could have been modernized with formatted string literals (f”” strings) when we upgraded to Python 3.7 in late 2019 (my statement also hints at the work to be done once NV Access and contributors upgrade to a newer Python 3 release in the future as there are code that could see some modernization love). Thanks.

michaelDCurran committed:
Revert "Improve responsiveness of input and focus by pumping immediately instead of after a delay. (#14701)" (#14925)

This reverts pr #14701, commit cc8b5ad

Link to issue number:
Fixes #14753
May fix #14803

Summary of the issue:
After the merging of pr #14701, It was reported in #14753 that with input help mode on and moving your finger around the touch screen, many errors would be logged and eventually input help mode would be disabled.

The recursion itself was caused by the core pump's Notify method calling request directly, which itself calls Notify.
The obvious fix would be to wx.CallAfter the call to request. However, this then caused the WX event loop to fill up with responses to these call afters, and due to the way the wx event loop works (msgWaitForMultipleObjects waiting on a pending event win32 event), the event loop never ended up getting a chance to process OS and 3rd party messages. And because we monkeypatch CallAfter to PostMessage (in case we are in a win32 modal or menu and don't wake up) the win32 message queue was being filled completely thus not allowing future messages to be posted.
@jcsteh and I have considered several different ways of trying to work around this. Most of which have not been successful. There are probably more things we could try, but for now it is better to revert until we can get this right.

Description of development approach
Fully reverts pr #14701 for now.

CyrilleB79 committed:
Report the state of cell formatting when using formatting shortcuts in Excel (#14923)

In Excel, when applying formatting to cells, nothing is reported. Instead, in Word, toggling bold, italic or underline is reported.

Description of user facing changes
When using Excel shortcuts to toggle the format of a full cell in Excel, the result is now reported. The following toggle are supported:
• bold: control+b or control+shift+2
• italic: control+I or control+shift+3
• underline: control+U or control+shift+4
• strikethrough: control+shift+5 (Note: did not find any other shortcut for it)

Description of development approach
• As done for Word text, use object model to look at the properties of the cell after the shortcut is executed.
• Copied the polling function of Word's code to check for value changes polling the object model. (I had initially thought that it was not required, but testing on another system, I have found that it was required)
• Implement the scripts on the whole worksheet so that the scripts work either on single cell or on multi-cell selections. Taking advantage that both excelCellObject and excelRangeObject have a font attribute.

codeofdusk commented on source/eventHandler.py:
```diff
@@ -294,19 +295,49 @@ def executeEvent(
+		log.exception(f"error executing event: {eventName} on {obj} with extra args of {kwargs}")
+
+
+virtualDesktopName: Optional[str] = None
```
This fails, because type hints can't be added to global names.

seanbudd replied:
type hints work fine for global names for me with VS Code - can you explain where this is coming from?
These are used regularly in NVDA

codeofdusk replied:
To clarify, the following exception is raised when I attempt to run this PR:

```python traceback
Traceback (most recent call last):
  File "nvda.pyw", line 406, in <module>
    core.main()
  File "core.py", line 510, in main
    import NVDAHelper
  File "NVDAHelper.py", line 27, in <module>
    import eventHandler
  File "eventHandler.py", line 324
    virtualDesktopName: Optional[str] = None
                                           ^
SyntaxError: annotated name 'virtualDesktopName' can't be global
```

seanbudd replied:
this is a scoping problem. If the line is moved higher in the file, above the first `global` declaration the code runs fine.

seanbudd replied:
>
Hi, I see. I’m sure we have other files with this issue, so I think it would be great to resolve these as part of 2024.1 development cycle. Thanks.

I think it's unlikely we have other files with this issue. This bug prevents NVDA from starting if affected.

josephsl commented:
Hi, this I didn’t know – type hint will be removed unless folks suggest alternatives. Thanks.

josephsl commented:
Hi, I see. I’m sure we have other files with this issue, so I think it would be great to resolve these as part of 2024.1 development cycle. Thanks.

josephsl committed:
Merge branch 'master' into i5641announceVirtualDesktops

josephsl committed:
Event handler: move virtual desktop constants to the top of the file.

Reported by several testers: in some cases, NVDA fails to start due to type hint scope issue if a constant is placed between functions. Therefore, move virtual desktops handling to top fo the file next to other constants and data structures.

josephsl commented:
Hi,

Try commit 0b5eb3a and see if it runs this time.

Also, the approach that Jamie did a while ago would have worked if his PR was adapted to use changes made to initial desktop object definition (core.py), but even then, it required editing event handler to handle cases where multiple desktops are opened in quick succession (pressing Control+Windows+D multiple times quickly as the focus will stay on desktop icons).

Thanks.

josephsl committed:
Event handler: lint (Flake8 ET126)

josephsl commented:
Hi,

I don't know why, but Appveyor is not building this branch even after force pushing commits.

Thanks.

seanbudd commented:
Hi @josephsl - strange. Could you consider opening a new PR so we can get a build?

codeofdusk commented:
Superseded by #14932?

seanbudd commented:
Closing in favour of https://github.com/nvaccess/nvda/pull/14932
